swim/member.go

```diff
@@ -73,6 +80,55 @@ func (m Member) checksumString(b *bytes.Buffer) {
 	b.WriteString(m.Address)
 	b.WriteString(m.Status)
 	b.WriteString(strconv.FormatInt(m.Incarnation, 10))
+	m.Labels.checksumString(b)
+}
+
+// checksumString adds the label portion of the checksum to the buffer that is
+// passed in. The string will not be appended in the case where labels are not
+// set on this member. This is for backwards compatibility reasons with older
+// versions.
+func (l LabelMap) checksumString(b *bytes.Buffer) {
+	checksum := l.checksum()
+
+	if checksum == 0 {
+		// we don't write the checksum of the labels if the value of the checksum
+		// is 0 (zero) to be backwards compatible with ringpop applications on
+		// an older version. This only works if the newer version does not use
+		// labels
+		return
+	}
+
+	// write #labels<checksum> to the buffer which will be appended to the
+	// checksum string for the node.
+	b.WriteString("#labels")
+	b.WriteString(strconv.Itoa(int(checksum)))
+}
+
+// checksum computes a checksum for the labels. It will return 0 (zero) when no
+// labels are set, but 0 does not indicate that no labels are set. It could be
+// possible that 0 is computed as the checksum.
+func (l LabelMap) checksum() int32 {
+	var lb bytes.Buffer
+	var checksum uint32
+	for key, value := range l {
+		// decide if we want to escape both fields to enforce uniqueness of the
+		// key-value checksum.
+		lb.WriteString("/")
+		lb.WriteString(key)
+		lb.WriteString("/")
+		lb.WriteString(value)
+
+		checksum = checksum ^ farm.Fingerprint32(lb.Bytes())
+
+		lb.Reset()
+	}
+
+	var signedChecksum int32
+	// This line converts an unsigned integer to a signed integer (32 bits).
+	// It is needed to be able to calculate the same value in nodejs for
+	// ringpop-node and the integration tests in ringpop-common.
```

Comment:
I wrote this because the other was not doable in nodejs. I needed to write the nodejs version because of the integration tests: https://github.com/uber/ringpop-common/blob/feature/labels/test/membership-checksum.js#L35-L42